### ISSUE TITLE
v1.5.1 & master: ImportError with vmaf python tools

### ISSUE THREAD

**Issue by dvrogozh** [Netflix/vmaf#541]
Please, clarify whether you officially support installing vmaf python module via python setuptools (https://github.com/Netflix/vmaf/blob/master/python/setup.py)? If you do, can you, please, document that in https://github.com/Netflix/vmaf/blob/master/resource/doc/VMAF_Python_library.md?

I observe "ImportError" using vmaf-1.5.1 python tools. For example:
```
# cd $(vmaf_cloned_copy)/python
# python3 setup.py build
# python3 setup.py install --user
# python3 -c 'import vmaf; print(vmaf.__license__);'
ImportError
ImportError
ImportError
ImportError
ImportError
ImportError
ImportError
BSD+Patent
# python3 --version
Python 3.6.8
```

Same issue happens if I directly follow setup instructions on https://github.com/Netflix/vmaf/blob/master/resource/doc/VMAF_Python_library.md (`export PYTHONPATH=$(pwd)/python/src`)

I believe that's regression comparing to v1.3.15 where I don't see this issue.

**Comment by slhck** [comment]
> Please, clarify whether you officially support installing vmaf python module via python setuptools

This is not supported at the moment. At least it was never intended to be installed that way.

> If you do, can you, please, document that (…)

The documentation should be updated soon.

> I observe "ImportError" using vmaf-1.5.1 python tools. For example:

Yes, there have been other reports of this as well (e.g. [here](https://github.com/Netflix/vmaf/issues/530#issuecomment-605325717)).

> I believe that's regression comparing to v1.3.15 where I don't see this issue.

That's interesting. I haven't yet checked from which files/packages these errors originate.

**Comment by dvrogozh** [comment]
>> I believe that's regression comparing to v1.3.15 where I don't see this issue.

> That's interesting. I haven't yet checked from which files/packages these errors originate.

Here we go. If I did not mistake:

```
$ git bisect start
$ git bisect bad v1.5.1
$ git bisect good v1.3.15

$ git bisect log
git bisect start
# bad: [35c6044d8483a3ab9528f1cc81f75738efe4f4c0] Merge pull request #509 from Netflix/feature/update-version-changelog-readme
git bisect bad 35c6044d8483a3ab9528f1cc81f75738efe4f4c0
# good: [e434247f8bf8f5ec36ba87062656583c073394fd] Feature/update doc (#361)
git bisect good e434247f8bf8f5ec36ba87062656583c073394fd
# good: [a6ff0cfbd7bbc2b7e4617decd6c31dac79307d47] Merge pull request #421 from Netflix/fix/put-tests-back
git bisect good a6ff0cfbd7bbc2b7e4617decd6c31dac79307d47
# bad: [fdf6089b3442e5283438d1ef3a853fbc72fb1498] libvmaf/output: log xml scores as %.6f
git bisect bad fdf6089b3442e5283438d1ef3a853fbc72fb1498
# good: [0858c55e9d509527f3be1aa09061be9ab0d3b80d] libvmaf: do away with VmafScore
git bisect good 0858c55e9d509527f3be1aa09061be9ab0d3b80d
# bad: [b002092b7d5cd207c6fe59585883fab540826070] libvmaf: always use *VmafModel in public api
git bisect bad b002092b7d5cd207c6fe59585883fab540826070
# good: [0c01dfdbbf594d3e12b5e9bab0b0be18d6c83c3c] libvmaf/vmaf_score_at_index: fix typo
git bisect good 0c01dfdbbf594d3e12b5e9bab0b0be18d6c83c3c
# good: [1b410fc4b3544772e865bce243e1b3c2cca759b7] libvmaf: implement vmaf_fex_float_vif
git bisect good 1b410fc4b3544772e865bce243e1b3c2cca759b7
# good: [ad45bfcbf664aeeff355b0ad63563251f8358232] libvmaf: implement vmaf_fex_float_motion
git bisect good ad45bfcbf664aeeff355b0ad63563251f8358232
# bad: [46c908c057ab8d5771c78021081d7dd164355f39] Exend ExternalProgram to allow passed in path.
git bisect bad 46c908c057ab8d5771c78021081d7dd164355f39
# first bad commit: [46c908c057ab8d5771c78021081d7dd164355f39] Exend ExternalProgram to allow passed in path.
```

So, this points out to 46c908c057ab8d5771c78021081d7dd164355f39.

I am not expert in python... but this line pays attention:
https://github.com/Netflix/vmaf/blob/c18a39d227b9e2fef317aee134501248f6f18087/python/src/vmaf/__init__.py#L22

**Comment by slhck** [comment]
You're right. This again calls the methods in `config.py`, which call `cls._path_from_external()`, which again does this:

https://github.com/Netflix/vmaf/blob/c18a39d227b9e2fef317aee134501248f6f18087/python/src/vmaf/config.py#L32

**Comment by dvrogozh** [comment]
We probably need a comment from @li-zhi as an author of 46c908c.

It seems this 'ImportError' is intended to warn/error out when external tools like ffmpeg are missing. But not all the tools depend on ffmpeg (for example bdrate calculator does not), so I guess this indeed should not be an error on the full module load. But maybe it makes sense to report a warning/error when someone uses tools which really rely on external tools? This makes me think that maybe ` from . import config ` itself should be refactored somehow rather than `ImportError` fully dropped? Or that's something you will fix going forward?

**Comment by slhck** [comment]
I agree it would make sense to refactor this entirely, and make the message more user-friendly -— and dependent on whether it is really an error or just a warning.

I wasn't going to address this part of the code (or the setuptools installation), just the docs for the time being. That being said I think it would be preferable to be able to install the library as you tried.

**Comment by dvrogozh** [comment]
Your change (https://github.com/slhck/vmaf/commit/eff05a5f9f7469ec9cbacfdab1a43456c7a50a90) works for me. No "ImportError".

> I wasn't going to address this part of the code (or the setuptools installation), just the docs for the time being.

That's just fine. Short term solution is entirely ok. I am just curious on the further plans.

> That being said I think it would be preferable to be able to install the library as you tried.

Well, this seems to actually work though I did not give it a thorough try. The part I am interesting into is metrics calculators, bdrate at the first place. There is another observation on my side about that by the way. Here is a a memory footprint estimation for vmaf and modules it depends on:
```
# python3 setup.py build
# python3 setup.py install --user
# du -sh ~/.local/
308M    ~/.local/
# cd ~/.local/lib/python3.6/site-packages/
# for d in `ls`; do du -sh $d; done
44K     cycler-0.10.0-py3.6.egg
48K     decorator-4.4.2-py3.6.egg
4.0K    easy-install.pth
8.2M    h5py-2.10.0-py3.6-linux-x86_64.egg
4.3M    imageio-2.8.0-py3.6.egg
1.3M    joblib-0.14.1-py3.6.egg
272K    kiwisolver-1.2.0-py3.6-linux-x86_64.egg
25M     matplotlib-3.2.1-py3.6-linux-x86_64.egg
6.2M    networkx-2.4-py3.6.egg
72M     numpy-1.18.2-py3.6-linux-x86_64.egg
33M     pandas-1.0.3-py3.6-linux-x86_64.egg
6.1M    Pillow-7.1.1-py3.6-linux-x86_64.egg
308K    pyparsing-3.0.0a1-py3.6.egg
500K    python_dateutil-2.8.1-py3.6.egg
2.8M    pytz-2019.3-py3.6.egg
8.1M    PyWavelets-1.1.1-py3.6-linux-x86_64.egg
40M     scikit_image-0.16.2-py3.6-linux-x86_64.egg
22M     scikit_learn-0.22.2.post1-py3.6-linux-x86_64.egg
79M     scipy-1.4.1-py3.6-linux-x86_64.egg
352K    sureal-0.4.2-py3.6.egg
1004K   vmaf-1.5.1-py3.6.egg
```
The above list of packages is entire list which got pulled on my system after `python3 setup.py install --user` (I explicitly deleted ~/.local before running it). As you see footprint is ~300MB and its actually a lot if you just need bdrate calculator from the vmaf toolset. This made me think whether there is a sense to have 2 vmaf python packages:
1. One light weight package with minimal dependencies. Metrics calculator and some scripts on top of ffmpeg might go into here
2. Another higher weight package. Here model training staff might go.

The 1st package would be useful for those who evaluates existing product quality (which often is accompanied by sample/demo docker containers where size is important). The 2nd package would be useful to be included into products.

Well, I am not sure whether such separation would actually make sense, but I thought I'll bring this to the attention.

**Comment by li-zhi** [comment]
@dvrogozh thanks for digging into this. This is indeed a bug I introduced a month ago or so. @slhck you can go ahead with the current PR (if ready I'll merge it) and I can work on a proper fix and bringing back the import warning.

@dvrogozh regarding the footprint issue: is the bdrate calculator the only thing you are interested in? If yes, it is entirely possible you just copy that file over (together with interpolation_utils). They have no dependency on other part of the repo.

**Comment by slhck** [comment]
@li-zhi Feel free to merge the PR for now, and I'll continue updating the docs. I must admit I don't know exactly how these externals are being used, so if I run into issues while trying it (from an end user perspective) I'll dig into the source code.

**Comment by slhck** [comment]
@dvrogozh Good that *in principle* it works, thanks for checking. We can explain how to install the package, but there should be a caveat about size.

**Comment by slhck** [comment]
I am in the middle of refactoring the package so it can be installed via setuptools/pip. Seems to work fine so far. I'll reference the PR here once it is available.

**Comment by dvrogozh** [comment]
> I'll reference the PR here once it is available.

Appreciate. I'll try it once it will be available. By the way, you might wish to support __version__ for vmaf package for the same trick you describe for your dependencies (https://github.com/Netflix/vmaf/blob/master/resource/doc/VMAF_Python_library.md#troubleshooting) to work:
```
# python3 -c 'import vmaf as pkg; print(pkg.__version__); print(pkg.__file__)'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
AttributeError: module 'vmaf' has no attribute '__version__'

```

**Comment by slhck** [comment]
Yeah, I've already considered this! We anyway need it for `setup.py`, as it shouldn't be pulled from a file outside of the `python` directory.

**Comment by dvrogozh** [comment]
Looking into dependencies specified in `setup.py` and `requirements.txt` I would question whether `scikit-learn` and `scikit-image` modules are actually needed since I don't see any `import` statements mentioning them:
```
# fgrep -rsn scikit python/
python/src/vmaf/core/train_test_model.py:959:        http://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestRegressor.html
python/src/vmaf/core/train_test_model.py:1000:        https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html
python/src/vmaf/core/train_test_model.py:1041:        http://scikit-learn.org/stable/modules/generated/sklearn.ensemble.ExtraTreesRegressor.html
python/src/vmaf.egg-info/requires.txt:5:scikit-learn>=0.18.1
python/src/vmaf.egg-info/requires.txt:6:scikit-image>=0.13.1
python/requirements.txt:5:scikit-learn>=0.18.1
python/requirements.txt:6:scikit-image>=0.13.1
python/setup.py:50:        "scikit-learn>=0.18.1",
python/setup.py:51:        "scikit-image>=0.13.1",
python/build/lib/vmaf/core/train_test_model.py:959:        http://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestRegressor.html
python/build/lib/vmaf/core/train_test_model.py:1000:        https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html
python/build/lib/vmaf/core/train_test_model.py:1041:        http://scikit-learn.org/stable/modules/generated/sklearn.ensemble.ExtraTreesRegressor.html
```

Maybe you use them in some optional way? Like optionally pointing to some model or files? In this case this might worth to write down in documents.

All other modules have `import` and some associated code.

**Comment by slhck** [comment]
`scikit-learn` uses `sklearn` as its package name, so you won't find it with `scikit` :wink: 

I've pushed what I have so far here: https://github.com/Netflix/vmaf/pull/544 – but it's not complete.

**Comment by NagabhushanSN95** [comment]
Hi,
I'm trying to get VMAF (v1.5.1) predictions on a set of videos. I'm getting this ImportError. Specifically in file `config.py`, line 26: `from . import externals`.

Is this some error that affects VMAF computation or am I safe?

**Comment by slhck** [comment]
This should now be fixed in `master`. Please check again.
